code/game/objects/items/devices/PDA/PDA.dm

```diff
@@ -904,6 +904,29 @@ GLOBAL_LIST_EMPTY(PDAs)
 			old_id.forceMove(get_turf(src))
 
 
+/obj/item/pda/pre_attack(obj/target, mob/living/user, params)
+	if(ismachinery(target))
+		var/obj/machinery/target_machine = target
+		if(!target_machine.panel_open && !istype(target, /obj/machinery/computer))
+			return ..()
+		if(!istype(cartridge, /obj/item/cartridge/virus/clown))
+			return ..()
+		var/obj/item/cartridge/virus/installed_cartridge = cartridge
+
+		if(installed_cartridge.charges <=0)
+			to_chat(user, "<span class='notice'>Out of charges.</span>")
+			return ..()
+		to_chat(user, "<span class='notice'>You upload the virus to the airlock controller!</span>")
+		var/sig_list
+		if(istype(target,/obj/machinery/door/airlock))
+			sig_list += list(COMSIG_AIRLOCK_OPEN, COMSIG_AIRLOCK_CLOSE)
+		else
+			sig_list += list(COMSIG_ATOM_ATTACK_HAND)
+			//TODO : Add more signals for specific cases like panel open, printing, dispensing etc
+		target.AddComponent(/datum/component/sound_player, amount = (rand(15,20)), signal_or_sig_list = sig_list)
+		installed_cartridge.charges --
+
```

Comment:
```suggestion
		target.AddComponent(/datum/component/sound_player, amount = (rand(15,20)), signal_or_sig_list = sig_list)
		installed_cartridge.charges --
		return TRUE

```
You'd want to stop the attack chain and not slap or open the airlock after this is done, correct?

Reply:
Yep ! It doesn't slap currently though if I'm not mistaken

Reply:
But I'll commit because Rohesie code good

Reply:
Doesn't it open it, if you have access?

Reply:
Errhh lemme check